src/NovelRT/Input/InteractionService.cpp

```diff
@@ -41,31 +57,43 @@ namespace NovelRT::Input {
       return;
     }
 
+    validateIfKeyCached(keyCode);
     processKeyState(keyCode, keyState);
   }
 
   void InteractionService::acceptMouseButtonClickPush(int button, int action, const Maths::GeoVector<float>& mousePosition) {
     auto keyState = static_cast<KeyState>(action);
     auto keyCode = static_cast<KeyCode>(button);
+    auto value = mousePosition.getVec4Value() * glm::scale(glm::vec3(1920.0f / _screenSize.getX(), 1080.0f / _screenSize.getY(), 0.0f));
 
-    _mousePositionsOnScreenPerButton.at(keyCode) = mousePosition.getVec4Value() * glm::scale(glm::vec3(1920.0f / _screenSize.getX(), 1080.0f / _screenSize.getY(), 0.0f));
+    auto result = _mousePositionsOnScreenPerButton.find(keyCode);
 
+    if (result == _mousePositionsOnScreenPerButton.end()) {
+      _mousePositionsOnScreenPerButton.insert({ keyCode, value });
+    }
+    else {
+      result->second = value;
+    }
+
+    validateIfKeyCached(keyCode);
     processKeyState(keyCode, keyState);
   }
 
   void InteractionService::HandleInteractionDraw(InteractionObject* target) {
     if (_keyStates[target->getSubscribedKey()] == KeyState::KeyDown
-      && target->validateInteractionPerimeter(_mousePositionsOnScreenPerButton[KeyCode::LeftMouseButton])
+      && target->validateInteractionPerimeter(_mousePositionsOnScreenPerButton[target->getSubscribedKey()])
       && (_clickTarget == nullptr || (_clickTarget->getLayer() > target->getLayer()))) {
-        _logger.logDebug("Valid click target detected! Executing...");
-        _clickTarget = target;
+      _logger.logDebug("Valid click target detected! Executing...");
+      _clickTarget = target;
     }
   }
 
-  void InteractionService::consumePlayerInput() {
-    processKeyState(KeyCode::LeftMouseButton, _keyStates.at(KeyCode::LeftMouseButton));
-    glfwPollEvents();
+void InteractionService::consumePlayerInput() {
+  for (auto it : _keyStates) {
```

Comment:
Isn't this going to process all the keystates that we have cached?

Reply:
It calls `tryProcessMouseState(KeyCode code)` which in turn checks if it's a mouse button.

However, I'll make it parameterless and loop through mouse buttons directly, that way it only loops for each mouse button instead of each registered key state which could get big.

Reply:
Right but the loop as-is will iterate 50 times even if the method doesn't do anything with it. Ty for the change. 😄